File: packages/typescript/test/publish.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { Agent } from "../src/index.js";
import { publish } from "../src/publish.js";

/** Build a real packed .aagent fixture so publish()'s verify() step passes. */
async function fixture(): Promise<string> {
  const dir = mkdtempSync(join(tmpdir(), "sdk-pub-"));
  const out = join(dir, "doer-0.1.0.aagent");
  await new Agent("essentials/doer", "0.1.0")
    .description("doer")
    .systemPrompt("You are the doer.")
    .model("claude-opus-4-7")
    .pack(out);
  return out;
}

test("publish derives artifact_type from kind and POSTs the tarball to /v1/submit", async () => {
  const out = await fixture();
  const original = globalThis.fetch;
  let captured: { url: string; artifactType: unknown; hasTarball: boolean; auth: string | null } | undefined;
  try {
    globalThis.fetch = (async (url: string | URL, init?: RequestInit) => {
      const form = init!.body as FormData;
      captured = {
        url: String(url),
        artifactType: form.get("artifact_type"),
        hasTarball: form.get("tarball") instanceof Blob,
        auth: new Headers(init!.headers).get("Authorization"),
      };
      return new Response(JSON.stringify({ assessment_id: "abc" }), { status: 202 });
    }) as typeof fetch;

    const r = await publish(out, { registry: "https://reg.example/", token: "sekret" });

    assert.equal(r.ok, true);
    assert.equal(r.status, 202);
    assert.equal(r.artifactType, "aagent");
    assert.equal(captured!.url, "https://reg.example/v1/submit"); // trailing slash trimmed
    assert.equal(captured!.artifactType, "aagent"); // derived from manifest kind
    assert.equal(captured!.hasTarball, true);
    assert.equal(captured!.auth, "Bearer sekret");
  } finally {
    globalThis.fetch = original;
  }
});

test("publish surfaces a non-2xx registry response as ok=false", async () => {
  const out = await fixture();
  const original = globalThis.fetch;
  try {
    globalThis.fetch = (async () =>
      new Response(JSON.stringify({ error: "stage1_kind_enum" }), { status: 400 })) as typeof fetch;
    const r = await publish(out, { registry: "https://reg.example" });
    assert.equal(r.ok, false);
    assert.equal(r.status, 400);
  } finally {
    globalThis.fetch = original;
  }
});

test("publish honors an explicit artifact_type override", async () => {
  const out = await fixture();
  const original = globalThis.fetch;
  let sentType: unknown;
  try {
    globalThis.fetch = (async (_url: string | URL, init?: RequestInit) => {
      sentType = (init!.body as FormData).get("artifact_type");
      return new Response("{}", { status: 202 });
    }) as typeof fetch;
    await publish(out, { registry: "https://reg.example", artifactType: "amodel:llm-backend" });
    assert.equal(sentType, "amodel:llm-backend");
  } finally {
    globalThis.fetch = original;
  }
});

File: packages/go/publish_test.go
package alexsdk_test

import (
	"bytes"
	"io"
	"net/http"
	"path/filepath"
	"strings"
	"testing"

	alexsdk "github.com/The10thHunter/alexandria-sdk/packages/go"
)

// mockDoer captures the request and returns a canned response — no live server.
type mockDoer struct {
	captured   *http.Request
	body       []byte
	statusCode int
	respBody   string
}

func (m *mockDoer) Do(req *http.Request) (*http.Response, error) {
	m.captured = req
	if req.Body != nil {
		m.body, _ = io.ReadAll(req.Body)
	}
	return &http.Response{
		StatusCode: m.statusCode,
		Body:       io.NopCloser(strings.NewReader(m.respBody)),
		Header:     make(http.Header),
	}, nil
}

func publishFixture(t *testing.T) string {
	t.Helper()
	dir := t.TempDir()
	out := filepath.Join(dir, "doer-0.1.0.aagent")
	if _, err := alexsdk.NewAgent("essentials/doer", "0.1.0").
		Description("doer").
		SystemPrompt("You are the doer.").
		Model("claude-opus-4-7").
		Pack(out); err != nil {
		t.Fatalf("pack fixture: %v", err)
	}
	return out
}

func TestPublishDerivesArtifactTypeAndPostsTarball(t *testing.T) {
	out := publishFixture(t)
	doer := &mockDoer{statusCode: 202, respBody: `{"assessment_id":"abc"}`}

	r, err := alexsdk.Publish(out, "https://reg.example/", alexsdk.PublishOptions{
		Token:      "sekret",
		HTTPClient: doer,
	})
	if err != nil {
		t.Fatalf("Publish: %v", err)
	}
	if !r.OK || r.Status != 202 {
		t.Fatalf("expected ok 202, got ok=%v status=%d", r.OK, r.Status)
	}
	if r.ArtifactType != "aagent" {
		t.Fatalf("expected artifact_type=aagent, got %q", r.ArtifactType)
	}
	if r.Name != "essentials/doer" {
		t.Fatalf("expected name=essentials/doer, got %q", r.Name)
	}
	if doer.captured.URL.String() != "https://reg.example/v1/submit" {
		t.Fatalf("expected trailing slash trimmed, got %q", doer.captured.URL.String())
	}
	if got := doer.captured.Header.Get("Authorization"); got != "Bearer sekret" {
		t.Fatalf("expected bearer auth, got %q", got)
	}
	if !strings.HasPrefix(doer.captured.Header.Get("Content-Type"), "multipart/form-data") {
		t.Fatalf("expected multipart content type, got %q", doer.captured.Header.Get("Content-Type"))
	}
	if !bytes.Contains(doer.body, []byte(`name="artifact_type"`)) || !bytes.Contains(doer.body, []byte("aagent")) {
		t.Fatal("multipart body missing artifact_type=aagent")
	}
	if !bytes.Contains(doer.body, []byte(`name="tarball"`)) || !bytes.Contains(doer.body, []byte(`filename="doer-0.1.0.aagent"`)) {
		t.Fatal("multipart body missing tarball part")
	}
}

func TestPublishSurfacesNon2xxAsNotOK(t *testing.T) {
	out := publishFixture(t)
	doer := &mockDoer{statusCode: 400, respBody: `{"error":"stage1_kind_enum"}`}

	r, err := alexsdk.Publish(out, "https://reg.example", alexsdk.PublishOptions{HTTPClient: doer})
	if err != nil {
		t.Fatalf("Publish: %v", err)
	}
	if r.OK {
		t.Fatal("expected ok=false for 400")
	}
	if r.Status != 400 {
		t.Fatalf("expected status 400, got %d", r.Status)
	}
}

func TestPublishHonorsArtifactTypeOverride(t *testing.T) {
	out := publishFixture(t)
	doer := &mockDoer{statusCode: 202, respBody: "{}"}

	r, err := alexsdk.Publish(out, "https://reg.example", alexsdk.PublishOptions{
		ArtifactType: "amodel:llm-backend",
		HTTPClient:   doer,
	})
	if err != nil {
		t.Fatalf("Publish: %v", err)
	}
	if r.ArtifactType != "amodel:llm-backend" {
		t.Fatalf("expected override artifact_type, got %q", r.ArtifactType)
	}
	if !bytes.Contains(doer.body, []byte("amodel:llm-backend")) {
		t.Fatal("multipart body missing overridden artifact_type")
	}
	// no auth header when no token
	if got := doer.captured.Header.Get("Authorization"); got != "" {
		t.Fatalf("expected no auth header, got %q", got)
	}
}
